File: programming_hw_4s/tasks_with_eolymp_tags/t_13_4_(eolimp_6130).py
class Node:

    def __init__(self, item):
        self.item = item
        self.next = None
        self.prev = None


class LinkedDequeue:

    def __init__(self):
        self._front = None
        self._back = None
        self._len = 0

    def push_front(self, item):
        node = Node(item)
        if self._front:
            self._front.prev = node
            node.next = self._front
        else:
            self._back = node
        self._front = node
        self._len += 1
        return 'ok'

    def push_back(self, item):
        node = Node(item)
        if self._back:
            self._back.next = node
            node.prev = self._back
        else:
            self._front = node
        self._back = node
        self._len += 1
        return 'ok'

    def pop_front(self):
        if self._front:
            node = self._front
            self._front = node.next
            if self._front:
                self._front.prev = None
            else:
                self._back = None
            self._len -= 1
            return node.item
        return 'error'

    def pop_back(self):
        if self._back:
            node = self._back
            self._back = node.prev
            if self._back:
                self._back.next = None
            else:
                self._front = None
            self._len -= 1
            return node.item
        return 'error'

    def front(self):
        return self._front.item if self._front else 'error'

    def back(self):
        return self._back.item if self._back else 'error'

    def size(self):
        return self._len

    def clear(self):
        self.__init__()
        return 'ok'

    def exit(self):
        return 'bye'

    def execute(self, command):
        command = command.split()
        name = command[0]
        args = command[1:]
        return getattr(self, name)(*args)


if __name__ == '__main__':
    new_linked_deque = LinkedDequeue()

    with open('input.txt') as input:
        with open('output.txt', 'w') as output:
            for line in input:
                result = new_linked_deque.execute(line.rstrip())
                print(result, file=output)
                
                if result == 'bye': break

File: programming_hw_4s/tasks_with_eolymp_tags/t_13_4_(eolimp_6130).cpp
#include <iostream>
#include <string>
using namespace std;

struct Node{
	int value;
	Node *next, *prev;
};

struct deque{
	Node *head=nullptr;
	Node *tail=nullptr;
	int count=0;
	void push_back(int num){
		Node* element=new Node;
		element->value=num;
		count++;
		if(!head){
			head=element;
			tail=head;
		}
		else{
			element->prev=tail;
			tail->next=element;
			tail=element;
		}
		cout<<"ok"<<endl;
	}
	void push_front(int num){
		Node *element=new Node;
		element->value=num;
		count++;
		if(!head){
			head=element;
			tail=head;
		}
		else{
			element->next=head;
			head->prev=element;
			head=element;
		}
		cout<<"ok"<<endl;
	}
	void pop_back(){
		if(count!=0){
			cout<<tail->value<<endl;
			if(count>1){
				Node *element=tail;
				tail=tail->prev;
				tail->next=nullptr;
				delete element;
				count--;
			}
			else{
				head=tail=0;
				count--;
			}
		}
		else cout<<"error"<<endl;
	}
	void pop_front(){
		if(count!=0){
			cout<<head->value<<endl;
			if(head->next){
				Node *element=head;
				head=head->next;
				head->prev=nullptr;
				delete element;
				count--;
			}
			else if(head==tail){
				head->next=nullptr;
				head=nullptr;
				delete head;
				count=0;
			}
		}
		else cout<<"error"<<endl;
	}
	void back(){
		if(count!=0)cout<<tail->value<<endl;
		else cout<<"error"<<endl;
	}
	void front(){
		if(count!=0)cout<<head->value<<endl;
		else cout<<"error"<<endl;
	}
	void size(){
		cout<<count<<endl;
	}
	void clear(){
		count=0;
		cout<<"ok"<<endl;
		while(head)
		{
			tail=head->next;
			delete head;
			head=tail;
		}
	}
	void exit(){
		cout<<"bye";
	}
};

int main() {
	deque deq;
	string str;
	while(cin>>str){
		if(str=="push_front"){
			int num;
			cin>>num;
			deq.push_front(num);
		}
		if(str=="push_back"){
			int num;
			cin>>num;
			deq.push_back(num);
		}
		if(str=="pop_front")deq.pop_front();
		if(str=="pop_back")deq.pop_back();
		if(str=="clear")deq.clear();
		if(str=="size")deq.size();
		if(str=="front")deq.front();
		if(str=="back")deq.back();
		if(str=="exit"){
			deq.exit();
			break;
		}
	}
	return 0;
}